### PR TITLE
[video_player] adding onComplete callback

### PR DESCRIPTION
**Pull request by omartinma**
PR to add onComplete callback when the video has finished playing. Following same changes were applied in this [PR](https://github.com/flutter/plugins/pull/2224) but it was never merged.

Fixes https://github.com/flutter/flutter/issues/21929

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

File: packages/video_player/video_player/CHANGELOG.md [hunk 1]
```diff
@@ -1,525 +1,527 @@
+## 2.1.16
+
+- Add on complete callback
+
 ## 2.1.15
 
-* Ensured seekTo isn't called before video player is initialized. Fixes [#89259](https://github.com/flutter/flutter/issues/89259).
```
Comment (stuartmorgan) [inline packages/video_player/video_player/CHANGELOG.md:3]: Please do not rewrite the entire CHANGELOG format.

**Comment by stuartmorgan** [comment]
Thanks for the contribution!

> [x] I listed at least one issue that this PR fixes in the description above.

Could you please list the issue that explains the use case for this API addition? You've checked the box here, but there's no issue link in the PR.

File: packages/video_player/video_player/CHANGELOG.md (hunk 1 above)
Comment (omartinma) [reply packages/video_player/video_player/CHANGELOG.md:3]: I just revert the change

**Comment by omartinma** [comment]
> Thanks for the contribution!
> 
> > [x] I listed at least one issue that this PR fixes in the description above.
> 
> Could you please list the issue that explains the use case for this API addition? You've checked the box here, but there's no issue link in the PR.

I linked in the description to a previous PR, I thought was enough. I will unmark this field then!

**Comment by stuartmorgan** [comment]
> > Thanks for the contribution!
> > > [x] I listed at least one issue that this PR fixes in the description above.
> > 
> > 
> > Could you please list the issue that explains the use case for this API addition? You've checked the box here, but there's no issue link in the PR.
> 
> I linked in the description to a previous PR, I thought was enough. I will unmark this field then!

The previous PR doesn't link to an issue either, or provide the context I was requesting.

**Comment by omartinma** [comment]
> > > Thanks for the contribution!
> > > > [x] I listed at least one issue that this PR fixes in the description above.
> > > 
> > > 
> > > Could you please list the issue that explains the use case for this API addition? You've checked the box here, but there's no issue link in the PR.
> > 
> > 
> > I linked in the description to a previous PR, I thought was enough. I will unmark this field then!
> 
> The previous PR doesn't link to an issue either, or provide the context I was requesting.

What should I do then?

**Comment by stuartmorgan** [comment]
Please file an issue (per [the contributing guide](https://github.com/flutter/flutter/wiki/Tree-hygiene#overview)) describing the use case for the new API you want to add, and link to it from the PR description.

**Comment by omartinma** [comment]
> Please file an issue (per [the contributing guide](https://github.com/flutter/flutter/wiki/Tree-hygiene#overview)) describing the use case for the new API you want to add, and link to it from the PR description.

I just created the [issue](https://github.com/flutter/flutter/issues/90030) @stuartmorgan . Hope is ok with that, thanks for helping!

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Based on the discussion in the two related issues, it seems like it's currently possible to determine this via the existing value listener, but it's not as easy as it could be.

I think a better approach would be to look at adding more information to the value to make it easier to determine that (e.g., maybe a "transition type" that could indicate special cases like started, ended, etc.), rather than adding a one-off, completely separate way of listening to one particular event.

**Comment by ditman** [comment]
I agree with @stuartmorgan in this one, the `onComplete` looks a little bit forced into the API. It is a shame that the VideoPlayerController doesn't expose a `Stream<VideoPlayerEvent>`, similar to the one listened to here:

https://github.com/flutter/plugins/blob/master/packages/video_player/video_player/lib/video_player.dart#L388-L389

Everything would be quite a bit simpler IMO with a Stream rather than with a ValueListenable + a complex value object (but I understand this is my personal taste, and probably not an absolute truth)

**Comment by omartinma** [comment]
Thanks both for you feedback! @stuartmorgan @ditman 
I like also the idea to expose `Stream<VideoPlayerEvent>` so it will give more flexibility, for instance maybe you want to know when it starts, paused, etc.
Should i update my PR or should we create new issue and create a new PR for that?

**Comment by stuartmorgan** [comment]
I think the next step would be to propose one or more new events in https://github.com/flutter/flutter/issues/21929 that would cover the use case, and we can discuss further there.

I'm going to close this pending a resolution of a discussion there. Thanks!

**Pull request closed without merging**